### PR TITLE
KAFKA-3018: added topic name validator to ProducerRecord

### PR DESCRIPTION
**Pull request by choang**
Added validation for topic name when creating a `ProducerRecord`, and added corresponding tests.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java [hunk 1]
```diff
@@ -56,6 +59,8 @@
     public ProducerRecord(String topic, Integer partition, Long timestamp, K key, V value) {
         if (topic == null)
             throw new IllegalArgumentException("Topic cannot be null");
+        if (!TopicValidator.isValidateTopicName(topic))
+            throw new InvalidTopicException(topic + " is not a valid topic name.");
```
Comment (granthenke) [inline clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java:63]: The logic for a valid topic name is actually a little more complicated than this. The server side validation exists in `kafka.common.Topic.scala`.

We also shouldn't duplicate the logic. So if we move it to the clients package, removing the Scala version and replacing its usages would be good too.

Also #898 (KAFKA-3219) may change this logic a little bit too.

Comment (choang) [inline clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java:63]: I originally wanted to use the logic from core's `Topic.scala`, however core isn't included as a dependency for clients.  Do you have any guidance on how to approach this?  I could see introducing a `common` module, and throwing in all the shared logic/code there, but that seems like a much broader conversation.

Comment (granthenke) [inline clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java:63]: My suggestion would be to move the logic from the core module `kafka.common.Topic.scala` to  the client module. And then replace all usages of the old Topic class in core with the new version added in clients.

`org.apache.kafka.common` is where we want all common client and server classes/logic to live.

Comment (choang) [inline clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java:63]: That seems like a good suggestion as core includes clients as a dependency.  @becketqin included a comment in the JIRA ticket that had a different preference.  @granthenke can you chime in on that?

**Comment by choang** [comment]
Per KAFKA-3018 discussion, we want to take a different approach to solving this problem, so I'm closing the PR in favor of the alternate design.

**Pull request closed without merging**